ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/core/Transaction.java

```diff
@@ -87,14 +89,14 @@ public static Transaction readFrom(final RLPInput input) throws RLPException {
             .value(input.readUInt256Scalar(Wei::wrap))
             .payload(input.readBytesValue());
 
-    final int v = input.readIntScalar();
+    final BigInteger v = input.readBigIntegerScalar();
     final byte recId;
-    int chainId = -1;
-    if (v == REPLAY_UNPROTECTED_V_BASE || v == REPLAY_UNPROTECTED_V_BASE + 1) {
-      recId = (byte) (v - REPLAY_UNPROTECTED_V_BASE);
-    } else if (v > REPLAY_PROTECTED_V_MIN) {
-      chainId = (v - REPLAY_PROTECTED_V_BASE) / 2;
-      recId = (byte) (v - (2 * chainId + REPLAY_PROTECTED_V_BASE));
+    BigInteger chainId = BigInteger.valueOf(-1);
+    if (v.equals(REPLAY_UNPROTECTED_V_BASE) || v.equals(REPLAY_UNPROTECTED_V_BASE_PLUS_1)) {
+      recId = v.subtract(REPLAY_UNPROTECTED_V_BASE).byteValue();
+    } else if (v.compareTo(REPLAY_PROTECTED_V_MIN) > 0) {
+      chainId = v.subtract(REPLAY_PROTECTED_V_BASE).divide(TWO);
+      recId = v.subtract(TWO.multiply(chainId).add(REPLAY_PROTECTED_V_BASE)).byteValue();
```

Comment:
Probably should use `byteValueExact` instead of `byteValue` here and line 96.

Reply:
done